File: sdk/src/types/permissions.ts
export type IPermissions = {
  mask: number;
};

export const Permission = {
  InitiateTransaction: 1 << 0,
  VoteTransaction: 1 << 1,
  ExecuteTransaction: 1 << 2,
  IsDelegate: 1 << 3,
} as const;

export type Permission = (typeof Permission)[keyof typeof Permission];

export class Permissions implements IPermissions {
  private constructor(readonly mask: number) {}

  static fromPermissions(permissions: Permission[]) {
    return new Permissions(
      permissions.reduce((mask, permission) => mask | permission, 0)
    );
  }

  static all() {
    return new Permissions(
      Object.values(Permission).reduce(
        (mask, permission) => mask | permission,
        0
      )
    );
  }

  static has(permissions: IPermissions, permission: Permission) {
    return (permissions.mask & permission) === permission;
  }
}

File: programs/multi_wallet/src/state/permissions.rs
use anchor_lang::prelude::*;
use bytemuck::{Pod, Zeroable};

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum Permission {
    InitiateTransaction = 1 << 0,
    VoteTransaction = 1 << 1,
    ExecuteTransaction = 1 << 2,
    IsPermanentMember = 1 << 3,
}

/// Bitmask for permissions.
#[derive(
    AnchorSerialize,
    AnchorDeserialize,
    InitSpace,
    PartialEq,
    Zeroable,
    Pod,
    Copy,
    Clone,
    Default,
    Debug,
)]
#[repr(C)]
pub struct Permissions {
    pub mask: u8,
}

impl Permissions {
    pub fn has(&self, permission: Permission) -> bool {
        self.mask & (permission as u8) != 0
    }

    pub fn from_permissions(permissions: Vec<Permission>) -> Self {
        let mask = permissions.iter().fold(0u8, |acc, f| acc | (*f as u8));

        Permissions { mask }
    }
}

#[derive(Default)]
pub struct PermissionCounts {
    pub voters: usize,
    pub initiators: usize,
    pub executors: usize,
    pub permanent_members: usize,
}
